### PR TITLE
Convert Rest of Com Vtbls to New Pattern

### PR DESCRIPTION
**Pull request by lonitra**
Follow up from #8311

- Convert `IDataSource`,`IDataSourceNotify`,`IDataObject`,`IEnumFORMATETC` to our new CCW pattern
- Add explicit conversion for `ComTypes.STGMEDIUM` and Cswin32 `STGMEDIUM`
- Implement the respective Cswin32 interfaces for our managed objects. 
   Note: 
    - `ComTypes.IDataObject` interface cannot be removed from `DataObject` due to it being a public class
    - `ComTypes.IEnumFORMATETC` interface cannot be removed from `FormatEnumerator` since it is used by `DataObject`
- Remove `EnumFORMATETCWrapper` 

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8330)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs [hunk 1]
```diff
@@ -1054,27 +1051,140 @@ public virtual void SetData(object? data)
             _innerData.SetData(data);
         }
 
-        unsafe ComInterfaceTable Com.IManagedWrapper.GetComInterfaceTable()
+        unsafe HRESULT Com.IDataObject.Interface.GetData(Com.FORMATETC* pformatetcIn, Com.STGMEDIUM* pmedium)
+        {
+            if (pmedium is null)
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs:1056]: what is the significance of `pmedium` check for null if this just being overridden on line 1064?

Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs:1056]: never mind. i see what is going.

**Review by dreddy-work — COMMENTED** [review]
LGTM

File: src/System.Windows.Forms/src/System/Windows/Forms/DropSource.cs [hunk 2]
```diff
@@ -3,148 +3,113 @@
...
-                    return HRESULT.DRAGDROP_S_CANCEL;
-                default:
-                    return HRESULT.S_OK;
-            }
+            action = DragAction.Cancel;
         }
-
-        public HRESULT GiveFeedback(Ole32.DROPEFFECT dwEffect)
+        else if (!grfKeyState.HasFlag(MODIFIERKEYS_FLAGS.MK_LBUTTON)
+                 && !grfKeyState.HasFlag(MODIFIERKEYS_FLAGS.MK_RBUTTON)
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/DropSource.cs:41]: nit: indenting

**Review by JeremyKuhne — CHANGES_REQUESTED** [review]
Nice!! A few questions, but mostly nits.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs [hunk 3]
```diff
@@ -1054,27 +1051,140 @@ public virtual void SetData(object? data)
...
+
+        unsafe HRESULT Com.IDataObject.Interface.DAdvise(Com.FORMATETC* pformatetc, uint advf, Com.IAdviseSink* pAdvSink, uint* pdwConnection)
+        {
+            var adviseSink = (IAdviseSink)Marshal.GetObjectForIUnknown((nint)(void*)pAdvSink);
+            return (HRESULT)((ComTypes.IDataObject)this).DAdvise(ref *(FORMATETC*)pformatetc, (ADVF)advf, adviseSink, out *(int*)pdwConnection);
+        }
+
+        HRESULT Com.IDataObject.Interface.DUnadvise(uint dwConnection)
+        {
+            try
```
[the remark this answers is not in the document]
Comment (lonitra) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataObject.cs:1154]: Ah that's right! We do wrap this at the CCW level. Thanks!

**Review by JeremyKuhne — APPROVED** [review]
There are still a few files that have no break between the class and the namespace. Other than that looks great!

**Review by JeremyKuhne — APPROVED** [review]
❤️‍🔥

**Pull request merged**